src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

```diff
@@ -261,15 +262,25 @@ public void write() {
               }
             } else {
               SinkRecord projectedRecord = SchemaUtils.project(record, currentSchema, compatibility);
-              writeRecord(projectedRecord);
-              buffer.poll();
-              if (shouldRotate(now)) {
+              String partition = partitioner.encodePartition(projectedRecord);
+              if (currentPartition != null && !currentPartition.equals(partition)) {
+                currentPartition = partition;
```

Comment:
I'm not sure I understand the goal of this logic. This task might be processing data for a lot of partitions, right? Won't this mean that if the framework delivers data for 3 topic partitions (say that it is conveniently split into 3 sets for each topic partition), we'll close both the first topic partition's and second topic partition's files by the time we get to the third?

Doesn't this violate the semantics that you want from the connector, which is that it tries to keep files open until enough data has been delivered into them? This is *really* important for HDFS since it often won't deal well with files that are, e.g., < 1GB. Of course it *can* handle smaller files, but to optimize performance, its better if you use larger files. The way we currently handle rolling files does this correctly. As far as I can tell, this change will result in a *ton* of tiny files given the way data is fetched from Kafka while your task is handling multiple partitions.

In particular, I think you are addressing the case of `TimeBasedPartitioner` specifically, where you are *also* assuming that timestamped data is delivered in order. That happens to be the case at the moment because of when this feature was added. But if `TimeBasedPartitioner` uses built-in Kafka timestamps, then its possible even within a topic partition, or across topic partitions (which this task might be consuming multiple of) we may see out of order data. For a general partitioner, the policy you're trying to use here doesn't make sense -- if you use `FieldPartitioner` there could be arbitrary ordering of records destined for different files.

I think this probably requires a slightly more complicated solution. One possibility would be to have the partitioner understand which partition files have "expired" and allow this code to handle that. I think that's actually a pretty major change as the `nextState()` call may depend on the current record. I think it may also require thinking through how we handle multi-file commit a bit more thoroughly.

@Ishiihara thoughts? You probably remember the semantics maintained across all these classes better than I do.